### PR TITLE
Rewrite cmake

### PR DESCRIPTION
**Pull request by Presburger**
1. rewrite CMakeLists.txt
2. modify some unnormal code
3. ut passed in aarch64 and mac os sillion.

### CODE REVIEW

File: knowhere/index/vector_index/IndexRHNSW.cpp [hunk 1]
```diff
@@ -34,7 +34,7 @@ IndexRHNSW::Serialize(const Config& config) {
         MemoryIOWriter writer;
         writer.name = this->index_type() + "_Index";
         faiss::write_index(index_.get(), &writer);
-        std::shared_ptr<uint8_t[]> data(writer.data_);
+        std::shared_ptr<uint8_t> data(writer.data_);
```
Comment (cydrain) [inline knowhere/index/vector_index/IndexRHNSW.cpp:37]: memory leak

File: knowhere/index/vector_index/IndexRHNSWPQ.cpp [hunk 2]
```diff
@@ -42,7 +42,7 @@ IndexRHNSWPQ::Serialize(const Config& config) {
             KNOWHERE_THROW_MSG("dynamic_cast<faiss::IndexRHNSWPQ*>(index_) failed during Serialize!");
         }
         faiss::write_index(real_idx->storage, &writer);
-        std::shared_ptr<uint8_t[]> data(writer.data_);
+        std::shared_ptr<uint8_t> data(writer.data_);
```
Comment (cydrain) [inline knowhere/index/vector_index/IndexRHNSWPQ.cpp:45]: memory leak

File: knowhere/index/vector_index/IndexRHNSWSQ.cpp [hunk 3]
```diff
@@ -45,7 +45,7 @@ IndexRHNSWSQ::Serialize(const Config& config) {
             KNOWHERE_THROW_MSG("dynamic_cast<faiss::IndexRHNSWSQ*>(index_) failed during Serialize!");
         }
         faiss::write_index(real_idx->storage, &writer);
-        std::shared_ptr<uint8_t[]> data(writer.data_);
+        std::shared_ptr<uint8_t> data(writer.data_);
```
Comment (cydrain) [inline knowhere/index/vector_index/IndexRHNSWSQ.cpp:48]: memory leak

File: knowhere/index/vector_index/IndexSPTAG.cpp [hunk 4]
```diff
@@ -59,19 +59,19 @@ CPUSPTAGRNG::Serialize(const Config& config) {
     snprintf(cstr, length, "%s", index_config.c_str());
 
     BinarySet binary_set;
-    std::shared_ptr<uint8_t[]> sample;
+    std::shared_ptr<uint8_t> sample;
```
Comment (cydrain) [inline knowhere/index/vector_index/IndexSPTAG.cpp:62]: memory leak

File: knowhere/index/vector_index/gpu/IndexGPUIDMAP.cpp [hunk 5]
```diff
@@ -48,7 +48,7 @@ GPUIDMAP::SerializeImpl(const IndexType& type) {
             faiss::write_index(host_index, &writer);
             delete host_index;
         }
-        std::shared_ptr<uint8_t[]> data(writer.data_);
+        std::shared_ptr<uint8_t> data(writer.data_);
```
Comment (cydrain) [inline knowhere/index/vector_index/gpu/IndexGPUIDMAP.cpp:51]: memory leak

**Review by cydrain — COMMENTED** [review]
too many unrelated code change, including:
1. coding format change
2. knowhere unittest directory change
3. thirdparty change to submodule (nlohmann/easylogging)
4. thirdparty code change (SPTAG/NSG)
5. faiss change (code format change, unittest target change)
6. unittest/xxx_benchmark removed (they are useful, why remove them ?)
7. ...

You'd better separate above items to multiple PRs for each.

File: .gitmodules [hunk 6]
```diff
@@ -0,0 +1,13 @@
+
+[submodule "thirdparty/cpu_features"]
+	path = thirdparty/cpu_features
+	url = https://github.com/google/cpu_features.git
+[submodule "thirdparty/nlohmann_json"]
```
Comment (cydrain) [inline .gitmodules:5]: why change nlohmann_json and easylogging to submodule and download them every time ?

File: CMakeLists.txt [hunk 7]
```diff
@@ -1,273 +1,229 @@
...
+
+list(REMOVE_ITEM KNOWHERE_SRCS ${KNOWHERE_SPTAG_SRCS})
+
+knowhere_file_glob(
+  GLOB FAISS_SRCS thirdparty/faiss/faiss/*.cpp
+  thirdparty/faiss/faiss/impl/*.cpp thirdparty/faiss/faiss/invlists/*.cpp
+  thirdparty/faiss/faiss/utils/*.cpp)
+
+knowhere_file_glob(GLOB FAISS_AVX512_SRCS
+                   thirdparty/faiss/faiss/impl/*avx512.cpp)
```
Comment (cydrain) [inline CMakeLists.txt:85]: there are *avx512.cpp files under thirdparty/faiss/faiss/utils

File: CMakeLists.txt [hunk 8]
```diff
@@ -1,273 +1,229 @@
...
+
+list(REMOVE_ITEM FAISS_SRCS ${FAISS_AVX512_SRCS})
+
+if(__AARCH64)
+knowhere_file_glob(GLOB FAISS_AVX_SRCS
+                   thirdparty/faiss/faiss/impl/*avx.cpp)
```
Comment (cydrain) [inline CMakeLists.txt:91]: there are *avx.cpp files under thirdparty/faiss/faiss/utils

File: CMakeLists.txt [hunk 9]
```diff
@@ -1,273 +1,229 @@
...
+  add_library(faiss_avx512 OBJECT ${FAISS_AVX512_SRCS})
+  target_compile_options(
+    faiss_avx512
+    PRIVATE $<$<COMPILE_LANGUAGE:CXX>:
+            -msse4.2
+            -mavx2
+            -mfma
+            -mf16c
+            -mavx512dq
+            -mavx512bw>)
```
Comment (cydrain) [inline CMakeLists.txt:111]: miss -mavx512f

File: CMakeLists.txt [hunk 10]
```diff
@@ -1,273 +1,229 @@
...
+)
+set(gtest_force_shared_crt
+    ON
+    CACHE BOOL "" FORCE)
+FetchContent_MakeAvailable(googletest)
+
+# knowhere_file_glob(GLOB KNOWHERE_UNITTEST_SRCS unittest/*.cpp )
+
+set(KNOWHERE_UNITTEST_SRCS
+    #unittest/test_annoy.cpp
```
Comment (cydrain) [inline CMakeLists.txt:173]: why disable test_annoy.cpp

File: CMakeLists.txt [hunk 11]
```diff
@@ -1,273 +1,229 @@
...
-                                --clang_format_binary   ${CLANG_FORMAT_BIN}
-                                --exclude_globs         ${LINT_EXCLUSIONS_FILE}
-                                --source_dir            ${CMAKE_CURRENT_SOURCE_DIR}/src
-                                ${KNOWHERE_LINT_QUIET} )
-endif ()
-
-#
-# "make clang-tidy" and "make check-clang-tidy" targets
-#
-if ( ${CLANG_TIDY_FOUND} )
```
Comment (cydrain) [inline CMakeLists.txt:229]: the original CMakeLists support clang_tidy, seems this function has been removed.

File: CMakeLists.txt [hunk 12]
```diff
@@ -1,273 +1,229 @@
...
-                            --exclude_globs     ${LINT_EXCLUSIONS_FILE}
-                            --source_dir        ${CMAKE_CURRENT_SOURCE_DIR}
-                            ${KNOWHERE_LINT_QUIET}
-                            )
-
-#
-# "make clang-format" and "make check-clang-format" targets
-#
-if ( ${CLANG_FORMAT_FOUND} )
-    # runs clang format and updates files in place.
```
Comment (cydrain) [inline CMakeLists.txt:208]: the original CMakeLists support clang_format, seems this function has been removed.

File: CMakeLists.txt [hunk 13]
```diff
@@ -1,273 +1,229 @@
...
+  add_library(knowhere_utils OBJECT ${KNOWHERE_UTILS_SRCS})
+
+  target_compile_options(
+    knowhere_utils
+    PRIVATE $<$<COMPILE_LANGUAGE:CXX>:
+            -msse4.2
+            -mavx2
+            -mfma
+            -mf16c
+            -mavx512dq
```
Comment (cydrain) [inline CMakeLists.txt:122]: miss -mavx512f

File: unittest/faiss_benchmark/faiss_benchmark_test.cpp [hunk 14]
```diff
@@ -1,593 +0,0 @@
-// Copyright (C) 2019-2020 Zilliz. All rights reserved.
```
Comment (cydrain) [inline unittest/faiss_benchmark/faiss_benchmark_test.cpp:1]: don't remove this directory

File: unittest/metric_alg_benchmark/metric_benchmark_test.cpp [hunk 15]
```diff
@@ -1,421 +0,0 @@
-// Copyright (C) 2019-2020 Zilliz. All rights reserved.
```
Comment (cydrain) [inline unittest/metric_alg_benchmark/metric_benchmark_test.cpp:1]: don't remove this directory

File: .gitmodules (hunk 6 above)
Comment (Presburger) [reply .gitmodules:5]: submodule is normal way in opensource project.

File: CMakeLists.txt (hunk 9 above)
Comment (Presburger) [reply CMakeLists.txt:111]: no need avx512f.

File: CMakeLists.txt (hunk 11 above)
Comment (Presburger) [reply CMakeLists.txt:229]: next time, we will use git hook to check this.

File: CMakeLists.txt (hunk 12 above)
Comment (Presburger) [reply CMakeLists.txt:208]: next time, we will use git hook to check this.

File: CMakeLists.txt [hunk 16]
```diff
@@ -1,273 +1,229 @@
...
-endif ()
-
-find_program( CPPLINT_BIN NAMES cpplint cpplint.py HINTS ${BUILD_SUPPORT_DIR} )
-message( STATUS "Found cpplint executable at ${CPPLINT_BIN}" )
-
-#
-# "make lint-knowhere" targets
-#
-add_custom_target( lint-knowhere
-    ${PYTHON_EXECUTABLE}    ${BUILD_SUPPORT_DIR}/run_cpplint.py
```
[the remark this answers is not in the document]
Comment (Presburger) [reply CMakeLists.txt:197]: next time, we will use git hook to check this.

File: CMakeLists.txt (hunk 13 above)
Comment (Presburger) [reply CMakeLists.txt:122]: no need.

File: thirdparty/faiss/faiss/FaissHook.cpp [hunk 17]
```diff
@@ -1,29 +1,66 @@
...
+#include <iostream>
+#include <mutex>
+#if defined(__x86_64__)
 #include <faiss/impl/ScalarQuantizerDC_avx.h>
 #include <faiss/impl/ScalarQuantizerDC_avx512.h>
-
+#endif
 namespace faiss {
 
+uint8_t lookup8bit[256] = {
```
[the remark this answers is not in the document]
Comment (Presburger) [reply thirdparty/faiss/faiss/FaissHook.cpp:15]: symbol conflict when use apple clang compiler.

**Pull request closed without merging**